# Recomanador Simple

## 1. INFORMACIÓ 

### 1.1. Abans de començar...

**\+ Durant la pràctica, solament es podran fer servir les següents llibreries**:

`Pandas, Numpy, Itertools`

*Nota: A més de les que ja es troben presents en la 1a cel·la i funcions natives de Python*

**\+ No es poden modificar les definicions de les funcions donades, ni canviar els noms de les variables i paràmetres ja donats**

Això no implica però que els hàgiu de fer servir. És a dir, que la funció tingui un paràmetre anomenat `df` no implica que l'hàgiu de fer servir, si no ho trobeu convenient.

**\+ En les funcions, s'especifica que serà i de quin tipus cada un dels paràmetres, cal respectar-ho**

Per exemple (ho posarà en el pydoc de la funció), `df` sempre serà indicatiu del `Pandas.DataFrame` de les dades. Durant la correcció, els paràmetres (i específicament `df`) no contindran les mateixes dades que en aquest notebook, si bé si seran del mateix tipus! Per tant, no us refieu de què tinguin, per exemple, el mateix nombre de files.

### 1.2. Consum general

La base de dades [movielens-1M](http://www.grouplens.org/node/73) conté 1,000,209 puntuacions de 3.900 pel·licules fetes l'any 2000 per 6.040 usuaris anònims del recomanador online [MovieLens](http://www.movielens.org/). 

El consum total de tots els usuaris s'hi pot trobar al document "ratings.dat" el format següent:

    UserID::MovieID::Rating::Timestamp

- **UserIDs** usuaris amb id's entre 1 i 6040 
- **MovieIDs** pelis amb id's entre 1 i 3952
- **Ratings** són les puntuacions en una escala de 5 estrelles.
- **Timestamp** representat en segons

> Cada usuari té com a mínim 20 interaccions consumides.

### 1.3. Usuaris



Al fitxer "users.dat" hi trobem la informació referent a cadascun dels usuaris en el següent format:

        UserID::Gender::Age::Occupation::Zip-code

- **Gender** ve donat per "M" per home i "F" per dona.
- **Age** està representada de la següent forma:

	*  1:  "Under 18"
	* 18:  "18-24"
	* 25:  "25-34"
	* 35:  "35-44"
	* 45:  "45-49"
	* 50:  "50-55"
	* 56:  "56+"

- **Occupation** es tria entre les següents opcions:

	*  0:  "other" or not specified
	*  1:  "academic/educator"
	*  2:  "artist"
	*  3:  "clerical/admin"
	*  4:  "college/grad student"
	*  5:  "customer service"
	*  6:  "doctor/health care"
	*  7:  "executive/managerial"
	*  8:  "farmer"
	*  9:  "homemaker"
	* 10:  "K-12 student"
	* 11:  "lawyer"
	* 12:  "programmer"
	* 13:  "retired"
	* 14:  "sales/marketing"
	* 15:  "scientist"
	* 16:  "self-employed"
	* 17:  "technician/engineer"
	* 18:  "tradesman/craftsman"
	* 19:  "unemployed"
	* 20:  "writer"

> Els usuaris han donat la informació voluntariament. Així doncs, alguns usuaris poden no tenir informació.


### 1.4. Películes



Al fitxer "movies.dat" hi trobem la informació referent a cadascuna de les películes en el següent format:

        MovieID::Title::Genres

- **Titles** són identics als titols de la base de dades IMDB, incloent l'any de llançament.
- **Genres** de les películes estan separats i seleccionats d'entre els següents:

	* Action
	* Adventure
	* Animation
	* Children's
	* Comedy
	* Crime
	* Documentary
	* Drama
	* Fantasy
	* Film-Noir
	* Horror
	* Musical
	* Mystery
	* Romance
	* Sci-Fi
	* Thriller
	* War
	* Western

> Algunes películes poden tenir el ID malament degut a duplicats accidentals.
>
> Les películes s'han entrat manualment, així que altres inconsistencies poden existir. 

## 2. COMENCEM: explorem les dades!!

### 2.1. Descarregar i llegir dades

+ Baixa't els fitxers que composen la base de dades i els còpies al teu directori de treball. 

+ Llegeix les tres taules de la base de dades en tres DataFrames de pandas amb aquest codi:

In [1]:
!wget https://files.grouplens.org/datasets/movielens/ml-1m.zip
!unzip ml-1m.zip

"wget" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.
"unzip" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.


In [2]:
import pandas as pd
unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_table('ml-1m/users.dat', sep='::', header=None, names=unames, engine='python')
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table('ml-1m/ratings.dat', sep='::', header=None, names=rnames, engine='python')
mnames = ['movie_id', 'title', 'genres']
movies = pd.read_table('ml-1m/movies.dat', sep='::', header=None, names=mnames, engine='python', encoding='latin-1')


In [3]:
print(type(unames))

<class 'list'>


### 2.2. Inspecció de les taules

In [4]:
print(len(users))
print(len(ratings))
print(len(movies))

6040
1000209
3883


In [5]:
users[:10]

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455
5        6      F   50           9  55117
6        7      M   35           1  06810
7        8      M   25          12  11413
8        9      M   25          17  61614
9       10      F   35           1  95370

In [6]:
users[-10:]

user_id gender  age  occupation    zip
6030     6031      F   18           0  45123
6031     6032      M   45           7  55108
6032     6033      M   50          13  78232
6033     6034      M   25          14  94117
6034     6035      F   25           1  78734
6035     6036      F   25          15  32603
6036     6037      F   45           1  76006
6037     6038      F   56           1  14706
6038     6039      F   45           0  01060
6039     6040      M   25           6  11106

In [7]:
ratings[-10:]

user_id  movie_id  rating  timestamp
1000199     6040      2022       5  956716207
1000200     6040      2028       5  956704519
1000201     6040      1080       4  957717322
1000202     6040      1089       4  956704996
1000203     6040      1090       3  956715518
1000204     6040      1091       1  956716541
1000205     6040      1094       5  956704887
1000206     6040       562       5  956704746
1000207     6040      1096       4  956715648
1000208     6040      1097       4  956715569

In [8]:
ratings[:10]

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291
5        1      1197       3  978302268
6        1      1287       5  978302039
7        1      2804       5  978300719
8        1       594       4  978302268
9        1       919       4  978301368

In [9]:
# ordena les puntuacions segons el 'movie_id'
ratings.sort_values('movie_id')[:5]
ratings.sort_values('rating')[-5:]

user_id  movie_id  rating  timestamp
760011     4516       296       5  964856871
321543     1906      2176       5  974690515
760009     4516       924       5  964856468
760487     4518      3429       5  964843900
0             1      1193       5  978300760

In [10]:
# les primeres 5 películes
movies[:5]

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

In [11]:
ratings[:5]

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

### **Exemple:** Fent càlculs sobre DataFrames.

Suposa que volem calcular les **puntuacions mitjanes d'una pel·licula per sexe o edat**. 


El primer pas a obtenir una única estructura que contingui tota la informació. Per fer-ho podem usar la funció ``merge`` de pandas. Aquesta funció infereix automàticament quines columnes ha d'usar per fer el ``merge`` basant-se en els noms que fan intersecció:

In [12]:
data = pd.merge(pd.merge(ratings, users), movies)

# Visualitzem la taula ordenada per identificar d'usuari
#data.sort_values(by='user_id')[:10]
data[:10]

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0        1      1193       5  978300760      F    1          10  48067   
1        2      1193       5  978298413      M   56          16  70072   
2       12      1193       4  978220179      M   25          12  32793   
3       15      1193       4  978199279      M   25           7  22903   
4       17      1193       5  978158471      M   50           1  95350   
5       18      1193       4  978156168      F   18           3  95825   
6       19      1193       5  982730936      M    1          10  48073   
7       24      1193       5  978136709      F   25           7  10023   
8       28      1193       3  978125194      F   25           1  14607   
9       33      1193       5  978557765      M   45           3  55421   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  
5  One Flew Over the Cuckoo's Nest (1975)  Drama  
6  One Flew Over the Cuckoo's Nest (1975)  Drama  
7  One Flew Over the Cuckoo's Nest (1975)  Drama  
8  One Flew Over the Cuckoo's Nest (1975)  Drama  
9  One Flew Over the Cuckoo's Nest (1975)  Drama

La funció ``iloc`` ens permet obtenir un subconjunt de files i/o columnes:

In [13]:
data.iloc[0:10]

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0        1      1193       5  978300760      F    1          10  48067   
1        2      1193       5  978298413      M   56          16  70072   
2       12      1193       4  978220179      M   25          12  32793   
3       15      1193       4  978199279      M   25           7  22903   
4       17      1193       5  978158471      M   50           1  95350   
5       18      1193       4  978156168      F   18           3  95825   
6       19      1193       5  982730936      M    1          10  48073   
7       24      1193       5  978136709      F   25           7  10023   
8       28      1193       3  978125194      F   25           1  14607   
9       33      1193       5  978557765      M   45           3  55421   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  
5  One Flew Over the Cuckoo's Nest (1975)  Drama  
6  One Flew Over the Cuckoo's Nest (1975)  Drama  
7  One Flew Over the Cuckoo's Nest (1975)  Drama  
8  One Flew Over the Cuckoo's Nest (1975)  Drama  
9  One Flew Over the Cuckoo's Nest (1975)  Drama

Els índexs Boolans ens permeten seleccionar una part de la taula que compleix una condició.

In [14]:
# comptem quin tant per cent de ratings estan fets per una dona

print(data['rating'].count())
print(float(len(data)))
print()

print(data[data['gender']=='F']['rating'].count())
print(type(data[data['gender']=='F']))
print(data[data['gender']=='F']['rating'].size)
print(data[data['gender']=='F']['rating'].count()/float(data['rating'].count()), '%')

1000209
1000209.0

246440
<class 'pandas.core.frame.DataFrame'>
246440
0.24638850480249627 %


Per obtenir les **puntuacions mitjanes de cada pel·licula agrupada per edat** podem usar el mètode ``pivot_table`` que és una forma de "canviar" la forma de la taula especificant quin valor agregat (mitjançant una funció predefinida) hi volem en funció dels valors de dues columnes:

In [15]:
print(type(data))
data[:5]

<class 'pandas.core.frame.DataFrame'>


user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0        1      1193       5  978300760      F    1          10  48067   
1        2      1193       5  978298413      M   56          16  70072   
2       12      1193       4  978220179      M   25          12  32793   
3       15      1193       4  978199279      M   25           7  22903   
4       17      1193       5  978158471      M   50           1  95350   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

In [16]:
mean_ratings = data.pivot_table(values= 'rating', index='title', columns='age', aggfunc='mean')
mean_ratings[:10]

age                                      1         18        25        35  \
title                                                                       
$1,000,000 Duck (1971)                  NaN  3.000000  3.090909  3.133333   
'Night Mother (1986)               2.000000  4.666667  3.423077  2.904762   
'Til There Was You (1997)          3.500000  2.500000  2.666667  2.900000   
'burbs, The (1989)                 4.500000  3.244444  2.652174  2.818182   
...And Justice for All (1979)      3.000000  3.428571  3.724138  3.657143   
1-900 (1994)                            NaN       NaN  2.000000       NaN   
10 Things I Hate About You (1999)  3.745455  3.415020  3.432950  3.102941   
101 Dalmatians (1961)              3.514286  3.295082  3.613757  3.826087   
101 Dalmatians (1996)              3.088235  2.467742  2.928571  3.279570   
12 Angry Men (1957)                4.176471  4.032609  4.408654  4.358333   

age                                      45        50        56  
title                                                            
$1,000,000 Duck (1971)             2.000000  2.750000       NaN  
'Night Mother (1986)               3.833333  3.555556  4.333333  
'Til There Was You (1997)          2.333333  2.500000  2.666667  
'burbs, The (1989)                 2.545455  3.208333  2.666667  
...And Justice for All (1979)      4.100000  3.551724  3.928571  
1-900 (1994)                            NaN       NaN  3.000000  
10 Things I Hate About You (1999)  3.258065  3.629630  4.000000  
101 Dalmatians (1961)              3.976744  3.650000  3.190476  
101 Dalmatians (1996)              3.482759  3.400000  3.555556  
12 Angry Men (1957)                4.274194  4.287879  4.235294

Per obtenir les **puntuacions mitjanes de cada pel·licula agrupada per sexe**:

In [17]:
mean_ratings = data.pivot_table('rating', index='title',columns='gender', aggfunc='mean')
mean_ratings[:10]

gender                                    F         M
title                                                
$1,000,000 Duck (1971)             3.375000  2.761905
'Night Mother (1986)               3.388889  3.352941
'Til There Was You (1997)          2.675676  2.733333
'burbs, The (1989)                 2.793478  2.962085
...And Justice for All (1979)      3.828571  3.689024
1-900 (1994)                       2.000000  3.000000
10 Things I Hate About You (1999)  3.646552  3.311966
101 Dalmatians (1961)              3.791444  3.500000
101 Dalmatians (1996)              3.240000  2.911215
12 Angry Men (1957)                4.184397  4.328421

Si volgéssim fer càlculs només sobre les pel·licules que han rebut **al menys** 250 puntuacions, primer hem de construir una taula amb el nombre d'avaluacions de cada títol. Per fer-ho, agruparem les dades per títol (amb el mètode ``groupby``) i usarem ``size()`` el nombre.

El mètode ``groupby`` implenta un o més d'aquests processos:

+ Dividir les dades segons algun criteri.
+ Aplicar una funció a cada grup.
+ Combinar els resultats en una estructura de dades.

In [18]:
ratings_by_title = data.groupby('title').size()
print(ratings_by_title)

title
$1,000,000 Duck (1971)                         37
'Night Mother (1986)                           70
'Til There Was You (1997)                      52
'burbs, The (1989)                            303
...And Justice for All (1979)                 199
                                             ... 
Zed & Two Noughts, A (1985)                    29
Zero Effect (1998)                            301
Zero Kelvin (Kjærlighetens kjøtere) (1995)      2
Zeus and Roxanne (1997)                        23
eXistenZ (1999)                               410
Length: 3706, dtype: int64


Llavors podem crear un índex amb els títols amb més de 250 avaluacions.

In [19]:
active_titles = ratings_by_title.index[ratings_by_title >= 250]
active_titles

Index([''burbs, The (1989)', '10 Things I Hate About You (1999)',
       '101 Dalmatians (1961)', '101 Dalmatians (1996)', '12 Angry Men (1957)',
       '13th Warrior, The (1999)', '2 Days in the Valley (1996)',
       '20,000 Leagues Under the Sea (1954)', '2001: A Space Odyssey (1968)',
       '2010 (1984)',
       ...
       'X-Men (2000)', 'Year of Living Dangerously (1982)',
       'Yellow Submarine (1968)', 'You've Got Mail (1998)',
       'Young Frankenstein (1974)', 'Young Guns (1988)',
       'Young Guns II (1990)', 'Young Sherlock Holmes (1985)',
       'Zero Effect (1998)', 'eXistenZ (1999)'],
      dtype='object', name='title', length=1216)

# PEL AMB ALMENYS 250 PUNTUACIONS-> mean-ratings


L'índex de títols que reben al menys 250 puntuacions es pot fer servir per seleccionar les files de ``mean_ratings``: 

In [20]:
mean_ratings = mean_ratings.loc[active_titles]
mean_ratings

gender                                    F         M
title                                                
'burbs, The (1989)                 2.793478  2.962085
10 Things I Hate About You (1999)  3.646552  3.311966
101 Dalmatians (1961)              3.791444  3.500000
101 Dalmatians (1996)              3.240000  2.911215
12 Angry Men (1957)                4.184397  4.328421
...                                     ...       ...
Young Guns (1988)                  3.371795  3.425620
Young Guns II (1990)               2.934783  2.904025
Young Sherlock Holmes (1985)       3.514706  3.363344
Zero Effect (1998)                 3.864407  3.723140
eXistenZ (1999)                    3.098592  3.289086

[1216 rows x 2 columns]

Per veure els films més valorats per les dones, podem ordenar per la columna F de forma descendent:

In [21]:
top_female_ratings = mean_ratings.sort_values(by='F', ascending=False)
top_female_ratings[:10]

gender                                                     F         M
title                                                                 
Close Shave, A (1995)                               4.644444  4.473795
Wrong Trousers, The (1993)                          4.588235  4.478261
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)       4.572650  4.464589
Wallace & Gromit: The Best of Aardman Animation...  4.563107  4.385075
Schindler's List (1993)                             4.562602  4.491415
Shawshank Redemption, The (1994)                    4.539075  4.560625
Grand Day Out, A (1992)                             4.537879  4.293255
To Kill a Mockingbird (1962)                        4.536667  4.372611
Creature Comforts (1990)                            4.513889  4.272277
Usual Suspects, The (1995)                          4.513317  4.518248

Suposem ara que volem les pel·licules que estan valorades de forma més diferent entre homes i dones. Una forma d'obtenir-ho és afegir una columna a ``mean_ratings`` que contingui la diferència en mitjana i llavors ordenar:

In [22]:
mean_ratings['diff'] = mean_ratings['M'] - mean_ratings['F']

Ordenant per ``diff`` ens dóna les pel·licules ben valorades per les dones que presenten més diferència entre homes i dones:

In [23]:
sorted_by_diff = mean_ratings.sort_values(by='diff')
sorted_by_diff[:15]

gender                                        F         M      diff
title                                                              
Dirty Dancing (1987)                   3.790378  2.959596 -0.830782
Jumpin' Jack Flash (1986)              3.254717  2.578358 -0.676359
Grease (1978)                          3.975265  3.367041 -0.608224
Little Women (1994)                    3.870588  3.321739 -0.548849
Steel Magnolias (1989)                 3.901734  3.365957 -0.535777
Anastasia (1997)                       3.800000  3.281609 -0.518391
Rocky Horror Picture Show, The (1975)  3.673016  3.160131 -0.512885
Color Purple, The (1985)               4.158192  3.659341 -0.498851
Age of Innocence, The (1993)           3.827068  3.339506 -0.487561
Free Willy (1993)                      2.921348  2.438776 -0.482573
French Kiss (1995)                     3.535714  3.056962 -0.478752
Little Shop of Horrors, The (1960)     3.650000  3.179688 -0.470312
Guys and Dolls (1955)                  4.051724  3.583333 -0.468391
Mary Poppins (1964)                    4.197740  3.730594 -0.467147
Patch Adams (1998)                     3.473282  3.008746 -0.464536

Invertint l'ordre de les files i fent un ``slicing`` de les 15 files superiors obtenim les pel·licules ben valorades pels homes que no han agradat a les dones: 

In [24]:
sorted_by_diff[::-1][:15]

gender                                         F         M      diff
title                                                               
Good, The Bad and The Ugly, The (1966)  3.494949  4.221300  0.726351
Kentucky Fried Movie, The (1977)        2.878788  3.555147  0.676359
Dumb & Dumber (1994)                    2.697987  3.336595  0.638608
Longest Day, The (1962)                 3.411765  4.031447  0.619682
Cable Guy, The (1996)                   2.250000  2.863787  0.613787
Evil Dead II (Dead By Dawn) (1987)      3.297297  3.909283  0.611985
Hidden, The (1987)                      3.137931  3.745098  0.607167
Rocky III (1982)                        2.361702  2.943503  0.581801
Caddyshack (1980)                       3.396135  3.969737  0.573602
For a Few Dollars More (1965)           3.409091  3.953795  0.544704
Porky's (1981)                          2.296875  2.836364  0.539489
Animal House (1978)                     3.628906  4.167192  0.538286
Exorcist, The (1973)                    3.537634  4.067239  0.529605
Fright Night (1985)                     2.973684  3.500000  0.526316
Barb Wire (1996)                        1.585366  2.100386  0.515020

Si volguéssim les pel·licules que han generat puntuacions més discordants, independentment del gènere, podem fer servir la variança o la desviació estàndard de les puntuacions: 

In [25]:
# Standard deviation of rating grouped by title
rating_std_by_title = data.groupby('title')['rating'].std()
# Filter down to active_titles
rating_std_by_title = rating_std_by_title.loc[active_titles]
rating_std_by_title.sort_values(ascending=False)[:10]

title
Dumb & Dumber (1994)                     1.321333
Blair Witch Project, The (1999)          1.316368
Natural Born Killers (1994)              1.307198
Tank Girl (1995)                         1.277695
Rocky Horror Picture Show, The (1975)    1.260177
Eyes Wide Shut (1999)                    1.259624
Evita (1996)                             1.253631
Billy Madison (1995)                     1.249970
Fear and Loathing in Las Vegas (1998)    1.246408
Bicentennial Man (1999)                  1.245533
Name: rating, dtype: float64

### Important: Temes de rendiment

In [26]:
%timeit data['title'] 
%timeit data.title 
%timeit data[['title']] 

3.34 µs ± 1.06 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)
12.3 µs ± 226 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
24 ms ± 1.68 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [27]:
type(data['title'])

pandas.core.series.Series

In [28]:
type(data[['title']])

pandas.core.frame.DataFrame

In [29]:
type(data.title)

pandas.core.series.Series

## 3. EXERCICIS

### 3.1. PRIMER EXERCICI 

Donada la taula ``data``, calcula la puntuació mitjana de cada usuari. 

In [30]:
!ls

assigment_code.ipynb
ml-1m


In [31]:
data_folder = 'ml-1m'

In [32]:
import pandas as pd
unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_table(f'{data_folder}/users.dat', sep='::', header=None, names=unames, engine='python')
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table(f'{data_folder}/ratings.dat', sep='::', header=None, names=rnames, engine='python')
mnames = ['movie_id', 'title', 'genres']
movies = pd.read_table(f'{data_folder}/movies.dat', sep='::', header=None, names=mnames, engine='python',encoding='latin-1')

data = pd.merge(pd.merge(ratings, users), movies)

In [33]:
data[:10]

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0        1      1193       5  978300760      F    1          10  48067   
1        2      1193       5  978298413      M   56          16  70072   
2       12      1193       4  978220179      M   25          12  32793   
3       15      1193       4  978199279      M   25           7  22903   
4       17      1193       5  978158471      M   50           1  95350   
5       18      1193       4  978156168      F   18           3  95825   
6       19      1193       5  982730936      M    1          10  48073   
7       24      1193       5  978136709      F   25           7  10023   
8       28      1193       3  978125194      F   25           1  14607   
9       33      1193       5  978557765      M   45           3  55421   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  
5  One Flew Over the Cuckoo's Nest (1975)  Drama  
6  One Flew Over the Cuckoo's Nest (1975)  Drama  
7  One Flew Over the Cuckoo's Nest (1975)  Drama  
8  One Flew Over the Cuckoo's Nest (1975)  Drama  
9  One Flew Over the Cuckoo's Nest (1975)  Drama

In [34]:
mean_user = data.groupby('user_id')['rating'].mean()
mean_user.to_frame()

rating
user_id          
1        4.188679
2        3.713178
3        3.901961
4        4.190476
5        3.146465
...           ...
6036     3.302928
6037     3.717822
6038     3.800000
6039     3.878049
6040     3.577713

[6040 rows x 1 columns]

+ Quina és la pel·lícula més ben puntuada (en mitja) pels usuaris? (Guarda aquest valor en una variable ``string``). 

In [35]:
best = data.groupby('title')['rating'].mean().sort_values().index[-1]
print(best)

Gate of Heavenly Peace, The (1995)


### 3.2. SEGON EXERCICI

Defineix una funció anomenada ``top_movie`` que donat un usuari ens retorni quina és la pel·lícula millor puntuada.<br> 


In [58]:
# la vostra solució aquí
def top_movie(dataFrame,usr):
    # obtenim la taula del l'usauri corresponent
    # ordenem aquesta taula segons les puntuacions
    # i agafem el millor puntuat
    # retornem l'id de la pel·lícula millor puntuada
    return dataFrame[dataFrame['user_id'] == usr].sort_values(by='rating')['title'][:1]
print(top_movie(data,1))

153545    Get Shorty (1995)
Name: title, dtype: object


### 3.3. TERCER EXERCICI

Construeix una funció que donat un dataframe et retorni el valor que cada usuari li ha donat a una peli. Això ho farem creant un dataframe on les columnes són els `movie_id`, les files `user_id` i els valors siguin el rating donat.

In [37]:
def build_rating_table(df):
    """
    Retorna un dataframe on les columnes són els `movie_id`, les files `user_id` i els valors
    la puntuació que li ha donat a la peli
    
    :param df: DataFrame original 
    :return: DataFrame descrit adalt
    """
    df = df.pivot_table(values = 'rating', index = 'user_id', columns = 'movie_id',fill_value = 0)
    return df

In [38]:
df_rating = build_rating_table(data)
df_rating

movie_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                               ...   
1            5     0     0     0     0     0     0     0     0     0  ...   
2            0     0     0     0     0     0     0     0     0     0  ...   
3            0     0     0     0     0     0     0     0     0     0  ...   
4            0     0     0     0     0     0     0     0     0     0  ...   
5            0     0     0     0     0     2     0     0     0     0  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
6036         0     0     0     2     0     3     0     0     0     0  ...   
6037         0     0     0     0     0     0     0     0     0     0  ...   
6038         0     0     0     0     0     0     0     0     0     0  ...   
6039         0     0     0     0     0     0     0     0     0     0  ...   
6040         3     0     0     0     0     0     0     0     0     0  ...   

movie_id  3943  3944  3945  3946  3947  3948  3949  3950  3951  3952  
user_id                                                               
1            0     0     0     0     0     0     0     0     0     0  
2            0     0     0     0     0     0     0     0     0     0  
3            0     0     0     0     0     0     0     0     0     0  
4            0     0     0     0     0     0     0     0     0     0  
5            0     0     0     0     0     0     0     0     0     0  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
6036         0     0     0     0     0     0     0     0     0     0  
6037         0     0     0     0     0     0     0     0     0     0  
6038         0     0     0     0     0     0     0     0     0     0  
6039         0     0     0     0     0     0     0     0     0     0  
6040         0     0     0     0     0     0     0     0     0     0  

[6040 rows x 3706 columns]

In [39]:
def build_counts_table(df):
    """
    Retorna un dataframe on les columnes són els `movie_id`, les files `user_id` i els valors
    el nombre de vegades que un usuari ha vist una peli d'un `movie_id`
    
    :param df: DataFrame original 
    :return: DataFrame descrit adalt
    """
    df = df.pivot_table(values = 'rating', index = 'user_id', columns = 'movie_id',fill_value = 0)
    df.replace({2:1,3:1,4:1,5:1},inplace = True)
    return df

In [40]:
df_counts = build_counts_table(data)
df_counts

movie_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                               ...   
1            1     0     0     0     0     0     0     0     0     0  ...   
2            0     0     0     0     0     0     0     0     0     0  ...   
3            0     0     0     0     0     0     0     0     0     0  ...   
4            0     0     0     0     0     0     0     0     0     0  ...   
5            0     0     0     0     0     1     0     0     0     0  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
6036         0     0     0     1     0     1     0     0     0     0  ...   
6037         0     0     0     0     0     0     0     0     0     0  ...   
6038         0     0     0     0     0     0     0     0     0     0  ...   
6039         0     0     0     0     0     0     0     0     0     0  ...   
6040         1     0     0     0     0     0     0     0     0     0  ...   

movie_id  3943  3944  3945  3946  3947  3948  3949  3950  3951  3952  
user_id                                                               
1            0     0     0     0     0     0     0     0     0     0  
2            0     0     0     0     0     0     0     0     0     0  
3            0     0     0     0     0     0     0     0     0     0  
4            0     0     0     0     0     0     0     0     0     0  
5            0     0     0     0     0     0     0     0     0     0  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
6036         0     0     0     0     0     0     0     0     0     0  
6037         0     0     0     0     0     0     0     0     0     0  
6038         0     0     0     0     0     0     0     0     0     0  
6039         0     0     0     0     0     0     0     0     0     0  
6040         0     0     0     0     0     0     0     0     0     0  

[6040 rows x 3706 columns]

També programarem una funció que donada la taula anterior i dos id's (usuari i peli), extregui el valor donat:

In [41]:
def get_count(df, user_id, movie_id):
    """
    Retorna el nombre de vegades que un usuari ha comprat en un `movie_id`
    
    :param df: DataFrame retornat per `build_counts_table`
    :param user_id: ID de l'usuari
    :param movie_id: ID de la peli
    :return: Enter amb el nombre de vegades que ha comprat
    """
    # tenir en compte que si user_id i movie_id comencen per 1, li hem de sumar 1 als dos indexos
    # un cop hàgim reindexat els ids dels usuaris i pelis no caldria sumar res
    return df[user_id+1][movie_id+1]

get_count(df_counts, 0, 0)

1

### 3.4. QUART EXERCICI

In [42]:
dataFrame = data
dataFrame.sort_values(by='movie_id')

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
43653      5881         1       4  957548112      M   35          17  30303   
42313      1988         1       4  974683131      F   25           1  85224   
42314      1991         1       4  974681894      M   25          20  48309   
42315      1992         1       3  974682228      M   18           4  85259   
42316      1993         1       4  974681718      M   35          17  99508   
...         ...       ...     ...        ...    ...  ...         ...    ...   
735529     1281      3952       2  974794160      M   25          19  97209   
735528     1270      3952       5  974814767      M   25           6  12230   
735527     1260      3952       4  974925406      M   25          17  28262   
735502     1079      3952       4  984260476      M   25           7  98115   
735403       36      3952       5  978062904      M   25           3  94123   

                        title                       genres  
43653        Toy Story (1995)  Animation|Children's|Comedy  
42313        Toy Story (1995)  Animation|Children's|Comedy  
42314        Toy Story (1995)  Animation|Children's|Comedy  
42315        Toy Story (1995)  Animation|Children's|Comedy  
42316        Toy Story (1995)  Animation|Children's|Comedy  
...                       ...                          ...  
735529  Contender, The (2000)               Drama|Thriller  
735528  Contender, The (2000)               Drama|Thriller  
735527  Contender, The (2000)               Drama|Thriller  
735502  Contender, The (2000)               Drama|Thriller  
735403  Contender, The (2000)               Drama|Thriller  

[1000209 rows x 10 columns]

In [43]:
data.nunique()

user_id         6040
movie_id        3706
rating             5
timestamp     458455
gender             2
age                7
occupation        21
zip             3439
title           3706
genres           301
dtype: int64

Si observem el nombre total d'usuaris unics i de pelicules úniques, podem observar que els id's dels usuaris van de 1 a 6040. Normalment hauriem d'intentar que comencessin al nombre 0, anant de 0 a 6039. 

**Què passa amb els indexos de les pelis??**

> ANSWER
>> Si llistem la taula data segons l'ordre dels ids de les pelis, podrem observar que l'id més gran que hi ha és el 3952, però amb la funció nunique només hem detectat 3706 ids, això es deu al fet que la funció ens retorna el nombre de pel·lícules diferents disponibles, que no és necessari que coincideixin aquests valors. Aquesta funció només ens dona la informació que hi ha 3706 pel·lícules diferents.

Usant la funció **pd.Categorical(*).codes**, re-indexa els id's dels usuaris i de les pelis perquè vagin de 0 a 6039 i de 0 a 3705 respectivament:

In [44]:
data['user_id'] = pd.Categorical(data['user_id']).codes
data['movie_id'] = pd.Categorical(data['movie_id']).codes
data.sort_values(by= 'movie_id')

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
43653      5880         0       4  957548112      M   35          17  30303   
42313      1987         0       4  974683131      F   25           1  85224   
42314      1990         0       4  974681894      M   25          20  48309   
42315      1991         0       3  974682228      M   18           4  85259   
42316      1992         0       4  974681718      M   35          17  99508   
...         ...       ...     ...        ...    ...  ...         ...    ...   
735529     1280      3705       2  974794160      M   25          19  97209   
735528     1269      3705       5  974814767      M   25           6  12230   
735527     1259      3705       4  974925406      M   25          17  28262   
735502     1078      3705       4  984260476      M   25           7  98115   
735403       35      3705       5  978062904      M   25           3  94123   

                        title                       genres  
43653        Toy Story (1995)  Animation|Children's|Comedy  
42313        Toy Story (1995)  Animation|Children's|Comedy  
42314        Toy Story (1995)  Animation|Children's|Comedy  
42315        Toy Story (1995)  Animation|Children's|Comedy  
42316        Toy Story (1995)  Animation|Children's|Comedy  
...                       ...                          ...  
735529  Contender, The (2000)               Drama|Thriller  
735528  Contender, The (2000)               Drama|Thriller  
735527  Contender, The (2000)               Drama|Thriller  
735502  Contender, The (2000)               Drama|Thriller  
735403  Contender, The (2000)               Drama|Thriller  

[1000209 rows x 10 columns]

Per comprovar que tot sigui correcte i guardar correctament la taula **df_counts**, torna a calcular-la:

In [45]:
df_counts = build_rating_table(data)
df_counts

movie_id  0     1     2     3     4     5     6     7     8     9     ...  \
user_id                                                               ...   
0            5     0     0     0     0     0     0     0     0     0  ...   
1            0     0     0     0     0     0     0     0     0     0  ...   
2            0     0     0     0     0     0     0     0     0     0  ...   
3            0     0     0     0     0     0     0     0     0     0  ...   
4            0     0     0     0     0     2     0     0     0     0  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
6035         0     0     0     2     0     3     0     0     0     0  ...   
6036         0     0     0     0     0     0     0     0     0     0  ...   
6037         0     0     0     0     0     0     0     0     0     0  ...   
6038         0     0     0     0     0     0     0     0     0     0  ...   
6039         3     0     0     0     0     0     0     0     0     0  ...   

movie_id  3696  3697  3698  3699  3700  3701  3702  3703  3704  3705  
user_id                                                               
0            0     0     0     0     0     0     0     0     0     0  
1            0     0     0     0     0     0     0     0     0     0  
2            0     0     0     0     0     0     0     0     0     0  
3            0     0     0     0     0     0     0     0     0     0  
4            0     0     0     0     0     0     0     0     0     0  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
6035         0     0     0     0     0     0     0     0     0     0  
6036         0     0     0     0     0     0     0     0     0     0  
6037         0     0     0     0     0     0     0     0     0     0  
6038         0     0     0     0     0     0     0     0     0     0  
6039         0     0     0     0     0     0     0     0     0     0  

[6040 rows x 3706 columns]

### 3.5. CINQUÉ EXERCICI



Construeix una funció <b>distEuclid(x,y)</b> que implementi la distància Euclidiana entre dos vectors usant funcions de pandas. Escriu la funció que calculin la semblança entre dos usuaris segons aquesta estructura:

``def SimEuclid (DataFrame, User1, User2)``
    Calcular els **embeddings** de cada usuari, C1 i C2, amb les puntuacions dels ítems comuns que han puntuat el dos usuaris.
    Si no hi ha puntuacions en comú, retornar 0. Retornar ``1/(1+distEuclid(C1, C2))``

\
Avalueu amb la funció ``%timeit`` quant triguen aquests càlculs per un parell d'usuaris.   

> *Nota: Alguns d'aquests exercicis tenen temps de càlcul de l'ordre de minuts sobre tota la base de dades. Per desenvolupar els algorismes és recomanable treballar amb una versió reduïda de la base de dades.* 

> *Nota: **embedding** fa referencia a un vector de N dimensions que té com a funció representar (en aquest cas) els gustos dels usuaris. També es pot fer un embedding d'items i aleshores representarien el contingut d'aquest o caracteristiques semblants.*

---
#### Mesurament de similituds (petita teoria)

El primer pas per poder recomanar és definir una funció de similitud entre vectors. Siguin $x, y$ vectors, implementa la seva distància euclidiana:

* Distància euclidea (inversa): https://en.wikipedia.org/wiki/Euclidean_distance

$$sim(x, y) = \frac{1}{1 + \sqrt{\sum_i(x_i-y_i)^2}}\in [0, 1]$$

---

Per implementar qualsevol d'aquestes únicament es permet l'ús de:

* `np.sum`
* `np.sqrt`
* `np.power`
* `np.dot`
* `np.linalg.norm`
* `np.mean`

I s'ha de fer **sense bucles**.


In [46]:
import math
import numpy as np
import pandas as pd

def distEuclid(x, y):
    """
    Retorna la distancia euclidiana de dos vectors n-dimensionals.
    
    :param x: Primer vector
    :param y: Segon vector
    :return : Escalar (float) corresponent a la distancia euclidiana
    """    
    dist = np.linalg.norm(x - y)
    return dist

def SimEuclid(DataFrame, User1, User2):
    """
    Retorna un score que representa la similitud entre user1 i user2 basada en la distancia euclidiana
    
    :param DataFrame: dataframe que conté totes les dades
    :param User1: id user1
    :param User2: id user2
    :return : Escalar (float) corresponent al score
    """
    # els vectors que farem servir seran les de la taula de ratings ordenats per usuari
    # ens assegurem que els dos parametres siguin vectors
    vector1 = DataFrame.iloc[User1] # Series
    vector2 = DataFrame.iloc[User2] 
    score = 1/(1+distEuclid(vector1, vector2))
    
    return score

In [47]:
# Execute functions
# millor obtenir la taula de les puntacions fora
rating_table =  build_rating_table(data)
%timeit SimEuclid(rating_table, 1, 5)

1.12 ms ± 224 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


### 3.6. SISÉ EXERCICI

Desenvolupa un sistema de recomanació col·laboratiu **basat en usuaris**. 

La funció principal, ``getRecommendationsUser``, ha de tenir com a entrada una taula de puntuacions, un ``user_id``, el tipus de mesura de semblança (Euclidiana) que volem usar, el nombre `m` d'usuaris semblants que volem per fer la recomanació i el nombre ``n`` de recomanacions que volem. Com a sortida ha de donar la llista de les ``n`` millors pel·lícules que li podriem recomanar segons la seva semblança amb altres usuaris.

> *Nota 1: S'ha d'evitar comparar ``user_id`` a ell mateix.*

> *Nota 2: Recordeu que en Python podem passar funcions com a paràmetres d'una funció.*

calcular la distancia euclidean en base dels ratings,
6. volem que cada usuari, calculem un diccionari de similituds que te amb la resta d'usuaris excepte a si mateix i cal normalitzar-la.

In [48]:
# getRecommendationsUser(data, 2, 50, 10, SimEuclid)

#### - SECCIÓ 1: 

Computa la score de similitud del usuari desitjat (userID) respecte tots els altres i retorna una llista dels **m** usuaris més propers, que seran els que usarem per fer la recomanació. Normalitzeu els scores de sortida.

In [49]:
dataFrame = build_rating_table(data)
dataFrame

movie_id  0     1     2     3     4     5     6     7     8     9     ...  \
user_id                                                               ...   
0            5     0     0     0     0     0     0     0     0     0  ...   
1            0     0     0     0     0     0     0     0     0     0  ...   
2            0     0     0     0     0     0     0     0     0     0  ...   
3            0     0     0     0     0     0     0     0     0     0  ...   
4            0     0     0     0     0     2     0     0     0     0  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
6035         0     0     0     2     0     3     0     0     0     0  ...   
6036         0     0     0     0     0     0     0     0     0     0  ...   
6037         0     0     0     0     0     0     0     0     0     0  ...   
6038         0     0     0     0     0     0     0     0     0     0  ...   
6039         3     0     0     0     0     0     0     0     0     0  ...   

movie_id  3696  3697  3698  3699  3700  3701  3702  3703  3704  3705  
user_id                                                               
0            0     0     0     0     0     0     0     0     0     0  
1            0     0     0     0     0     0     0     0     0     0  
2            0     0     0     0     0     0     0     0     0     0  
3            0     0     0     0     0     0     0     0     0     0  
4            0     0     0     0     0     0     0     0     0     0  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
6035         0     0     0     0     0     0     0     0     0     0  
6036         0     0     0     0     0     0     0     0     0     0  
6037         0     0     0     0     0     0     0     0     0     0  
6038         0     0     0     0     0     0     0     0     0     0  
6039         0     0     0     0     0     0     0     0     0     0  

[6040 rows x 3706 columns]

In [73]:
import itertools
def find_similar_users(DataFrame, userID, m, simfunction):
    """
    Retorna un diccionari de usuaris similars amb les scores corresponents.
    
    :param DataFrame: dataframe que conté totes les dades
    :param userID: usuari respecte al qual fem la recomanació
    :param m: nombre d'usuaris que volem per fer la recomanació
    :param similarity: mesura de similitud
    :return : diccionary
    """
    # primer obtenim la taula de les puntuacions dels usuaris
    scores = build_rating_table(DataFrame)
    
    # obtenim les similituds entre els usuaris
    #similarity = {(i if i != userId) : (simfunction(scores,userId,i)) for i in scores}
    similarity = dict()
    for i in scores:
        if i != userID:
            similarity[i] = simfunction(scores,userID,i)
    
    # sort del diccionari
    sort = sorted(similarity.items(), key=lambda item: item[1],reverse=True)
    sort 
    #out = dict(list(sort.items())[0: m])
    return sort

In [74]:
import datetime
t = datetime.datetime.now()
sim_dict = find_similar_users(data, 2, 10, SimEuclid)
t = datetime.datetime.now()-t
print(str(t))

<class 'list'>
0:00:05.775033


In [75]:
sim_dict

[(2922, 0.03346825709050096),
 (310, 0.03339095537904451),
 (3597, 0.03310582597772355),
 (2548, 0.03282795816068336),
 (2849, 0.03261068708575956),
 (3235, 0.032086035487345115),
 (3375, 0.03205196239416355),
 (1992, 0.03195039783542665),
 (1514, 0.03193356512249263),
 (383, 0.031916759277699894),
 (2835, 0.031866502236362934),
 (1762, 0.031833130575206814),
 (2713, 0.03178327122271373),
 (372, 0.031766703995965886),
 (3287, 0.03175016293213543),
 (3246, 0.03171715901712943),
 (2537, 0.03165146211931199),
 (3253, 0.03165146211931199),
 (3387, 0.03158617618052373),
 (159, 0.031569918404633975),
 (940, 0.031569918404633975),
 (276, 0.031537478838182774),
 (2127, 0.031489008463978195),
 (1492, 0.031424731311743816),
 (3211, 0.031424731311743816),
 (588, 0.031376783698388935),
 (2597, 0.031376783698388935),
 (316, 0.03131319727669042),
 (1528, 0.03129736163122709),
 (1090, 0.03126576305491067),
 (1430, 0.03126576305491067),
 (2685, 0.031234261018253537),
 (2798, 0.031234261018253537),
 (2

**CREIEU QUE ES OPTIM EL TEMPS QUE HA TARDAT PER UN USUARI?** RAONA LA RESPOSTA.

> ANSWER 
>> No és gaire òptim pel fet d'utilitzar fors per iterar la taula de puntuacions

Per millorar el problema anteorior, construeix una matriu de mida UXU on cada posició *i,j* indica la distància entre l'element *i* i el *j*. Així doncs, si estàs fent un recomanador basat en usuaris, `matriu[2, 3]` contindrà la similitud entre l'usuari 2 i el 3.

In [53]:
len(df_counts)
columns = df_counts.iloc[0]
columns

movie_id
0       5
1       0
2       0
3       0
4       0
       ..
3701    0
3702    0
3703    0
3704    0
3705    0
Name: 0, Length: 3706, dtype: int64

In [54]:
def similarity_matrix(similarity_function, df_counts):
    """
    Retorna una matriu de mida M x M on cada posició 
    indica la similitud entre usuaris (resp. ítems).
    
    :param similarity_function: Funció que calcularà la similitud 
        entre usuaris (resp. ítems)
    :param df_counts: Dataframe que conté el nombre de vegades que 
        un usuari ha comprat en un `aisle_id`
    :return : Matriu numpy de mida M x M amb les similituds.
    """
    
    # creem una nova matriu de M x M
    nUsers = len(df_counts)
    sim_matrix = np.zeros((nUsers, nUsers))
    
    # modifiquem els valors de la matriu on cada posició derà la similitud entre usuaris
            
    return sim_matrix

Per cridar aquesta funció, el primer paràmetre pot ser:

* Amb la funció de la distancia euclidiana que heu programat abans pot trigar ~@1h 30min treballant directament amb valors de numpy, ~@20h a partir de pandas pur.
* Opcionalment (no és obligatori fer-ho) podeu programar una funció que treballi específicament amb matrius (i no vectors). Si ho feu, cal gestionar-ho quan es rep `None`.  (@5s)

In [ ]:
%timeit sim = similarity_matrix(SimEuclid, df_counts)

In [ ]:
sim.shape

Ara torna a re-fer la funció **find_similar_users** i mira quant triga... 

> Recorda que les scores han d'estar normalitzades!

In [ ]:
def find_similar_users(DataFrame, sim_mx, userID, m):
    
    # la vostra solució aquí
    
    return dict(zip(users, w_scores))

In [ ]:
t = datetime.datetime.now()
sim_dict = find_similar_users(data, sim, 2, 10)
t = datetime.datetime.now()-t
print(str(t))

**Hauria d'haver baixat més de 30 cops el temps anterior!!**

#### - SECCIÓ 2: 
Computa les recomanacions per un usuari amb cadascun dels m usuaris més propers i fes una funció que retorni la **weighted average list** d'aquests per tal d'obtenir la recomanació final. Feu servir la funció anterior que usava la matriu de similituds per anar més ràpid!!


> *Nota: la **weighted average list** es calcularà agregant els n items més puntuats de cadascun dels m users més semblants al usuari donat.

In [ ]:
def getRecommendationsUser(DataFrame, user, sim_mx, n, m):
    """
    Retorna un dataframe de pel·licules amb els scores.
    
    :param DataFrame: dataframe que conté totes les dades
    :param user: usuari al qual fem la recomanació
    :param sim_mx: similarity_function
    :param n: nombre de pelis a recomanar
    :param m: nombre d'usuaris semblants a tenir en compte per les recomanacions
    :return : pandas de pelis amb els seus scores predits
    """
    # obtenim la fila d'auqest usuari 
    fila = DataFrame.iloc[user]
    
    #ordenem aquesta fila i obtenim els 5 usuaris més semblants
    DataFrame.sort_values(by=user, ascending=False, axis=1)
    users = [fila[i] for i in range(5)]
    
    
   
    
    return []

In [ ]:

t = datetime.datetime.now()
user_prediction = getRecommendationsUser(data, 3, sim, 10, 50)
t = datetime.datetime.now()-t
print(str(t))


In [ ]:
user_prediction

#### - SECCIÓ 3

Comprovem que realment l'usuari 3 no tenia recomanació per les pelis a les que estem recomanant:

In [ ]:
data_user3 = data[data['user_id']==3]
data_user3.head()

In [ ]:
desired_movie_id_to_check = 0
best_film_to_predict = user_prediction['movie_id'].iloc[desired_movie_id_to_check] 

In [ ]:
data_user3[data_user3['movie_id']==best_film_to_predict]

**Com es que ara estem predint un rating per la pelicula "best_film_to_predict"?** Digues quin nombre és la predicció

In [ ]:
# la vostra solució aquí

### 3.7. SETÉ EXERCICI


A continuació usarem la metrica Mean Absolute Error (MAE) per evaluar el nostre sistema. Aquesta mètrica ens permetrà mesurar la diferencia entre dues llistes donat un usuari: 

**1.** La llista amb els ratings originals d'un usuari donat

**2.** La llista de les prediccions generades per aquest usuari

#### - SECCIÓ 1: 

Treu el 10% dels usuaris i reserva aquests en una variable anomenada **test_set** i la resta en una variable anomenada **train_set**.

In [ ]:
import random
unique_users = data['user_id'].unique()
n_test_users = int(len(unique_users)*0.1)
n_train_users = len(unique_users) - n_test_users

test_usrs = random.sample(list(unique_users), n_test_users)

In [ ]:
n_test_users,n_train_users

In [ ]:
test_set = # la vostra solució aquí
train_set = # la vostra solució aquí
assert len(test_set) + len(train_set) == len(data)

Què passarà si calculo la matriu de similitud amb **train_set** i després intento predir pels usuaris de **test_set**??

#### - SECCIÓ 2:

Selecciona aproximadament el 80% de les interaccions de cada usuari de test i junta-les al **train_set**. Ara podem evaluar el sistema?


> Com la pràctica és molt llarga hem programat el codi per un usuari donat i vosaltres només heu de crear la funció que, per cada usuari, afagi el 80% de les intraccions i les unifiqui al dataframe de train.

In [ ]:
test_set.head()

In [ ]:
# Agafem el 20% de les pelis que ha consumit cada usuari de test 
groupby_count = test_set.groupby('user_id')['movie_id'].count()*0.2
groupby_count

Seleccionem la posició 1 i aquest use_id serà el que usarem pel codi d'exemple (que després haureu de replicar).

In [ ]:
groupby_count.reset_index().iloc[1]

In [ ]:
n_test_samples = int(groupby_count.reset_index().iloc[1]['movie_id'])
u = groupby_count.reset_index().iloc[1]['user_id']

In [ ]:
test_set_user = test_set[test_set['user_id'] == u]
frame_test = test_set_user.sample(n_test_samples)
print("TOTAL SAMPLES OF THE USER: " + str(len(test_set_user)))
print("TOTAL SAMPLES OF THE USER IN TEST SET: " + str(len(frame_test)))

In [ ]:
len(test_set_user.index)

In [ ]:
frame_train = test_set_user[~test_set_user.index.isin(frame_test.index)]
print("TOTAL SAMPLES OF THE USER IN TRAIN SET: " + str(len(frame_train)))

In [ ]:
assert len(frame_train) + len(frame_test) == len(test_set_user)

In [ ]:
def add_testdata(traindf, test_set):
    """
    Retorna els N usuaris més similars basat en la correlació de Pearson
    
    :param traindf: dataframe que conté les dades de train
    :param test_set: dataframe que conté les dades de test

    :return : 
        - :param 1st: dataframe que conté les dades de train juntament amb el 80% de test seleccionat
        - :param 2nd: dataframe que conté les dades de test que queden (20% restant)
    """
    
    # la vostra solució aquí
    
    return pd.concat((traindf, train_appended)), test_appended

In [ ]:
train, test = add_testdata(train_set, test_set)

In [ ]:
train.shape

In [ ]:
test.shape

#### -SECCIÓ 3:

Fes una funció usant **np.abs** que serveixi per evaluar el nostre sistema. **(PISTA:)** La funció ha de mesurar la diferencia en mitjana de la llista de ratings originals amb la llista de ratings predida.

In [ ]:
def evaluateRecommendations(train, test, m, n, sim):
    """
    Retorna l'error generat pel model
    
    :param DataFrame: dataframe que conté totes les dades
    :param userID: usuari respecte al qual fem la recomanació
    :param m: nombre d'usuaris que volem per fer la recomanació
    :param n: nombre de pelis a retornar
    :param sim: matriu de similitud
    :return : Escalar (float) corresponent al MAE
    """
    
    # la vostra solució 
        
    return 0

In [ ]:
t = datetime.datetime.now()
mae = evaluateRecommendations(train, test, 50, 10, sim)
t = datetime.datetime.now()-t
print(str(t))

In [ ]:
mae

### 3.8. VUITÉ (I ÚTLIM) EXERCICI


**Que surt més a compte, fer un recomanador unic pels dos sexes o un per cada sexe?** Justifica la resposta per escrit i amb el codi necessari.

> RESPOSTA

In [ ]:
# la vostra solució aquí